File: app/services/summary.py
"""Weekly aggregation: turn logged entries into per-muscle-group coverage.

The body map on the summary page shades a muscle group on a volume scale: from
light green at one set to dark green at the group's weekly target, then from
light red to dark red once the target is passed.  ``worked`` remains true from
the very first set, which is the rule the ``/log`` page and the breakdown list
key off.

Sets are **weighted by how directly a movement trains a group**: a primary
muscle takes the whole set, a secondary muscle half of it, so totals are
fractional (see :data:`app.exercises.PRIMARY_WEIGHT`). Movements outside
:data:`app.exercises.VOLUME_CATEGORIES` — stretches, cardio, plyometrics —
contribute nothing and never mark a group worked.

Six groups also break down into **regions** (see :data:`app.exercises.MUSCLE_REGIONS`).
Regions are graded differently — which is to say, not at all. They carry no
target, no state and no intensity, because no study has established how many
weekly sets a muscle *head* needs. They report where a group's volume landed and
whether a region was left out, which is a statement of fact rather than a
prescription. See docs/VOLUME_SCIENCE.md.
"""

from __future__ import annotations

from datetime import date

from ..exercises import (
    MUSCLE_GROUPS,
    MUSCLE_LABELS,
    REGION_LABELS,
    get_exercise,
    regions_for,
    regions_of,
)
from ..models import WorkoutEntry, list_entries
from ..training import DEFAULT_PROFILE, TrainerProfile
from .weeks import week_bounds, week_days


#: Share of a group's attributed volume below which a region reads as left out.
#:
#: **This is a judgement, not a finding**, and one of only two invented numbers
#: in the region feature. An even split across three delt heads is 33% and across
#: two chest regions 50%, so 15% is well clear of "merely less" in either case.
REGION_NEGLECT_SHARE = 0.15

#: A group must have had this much volume before a thin region means anything.
#:
#: Set to the approximate floor at which a muscle responds to training at all
#: (Pelland et al. 2025), so a group that was barely trained is reported as
#: barely trained rather than as three separate imbalances.
REGION_NEGLECT_MIN_PARENT_SETS = 4.0


def overshoot_span(target: int) -> int:
    """Sets past ``target`` that take a group from light red to dark red.

    Half the target, so an extra set is a visible step on either scale: ten
    sets of overshoot saturate a large group, five a small one.
    """
    return max(1, target // 2)


def grade(sets: float, target: int) -> tuple[str, float]:
    """Grade ``sets`` against ``target`` as a ``(state, intensity)`` pair.

    ``state`` is ``rest``, ``trained`` or ``over``; ``intensity`` runs 0–1
    *within that state's colour ramp*, so it restarts at the bottom when a
    group crosses its target and turns from green to red.

    ``sets`` is a float because secondary muscles count half (see the module
    docstring); the grading maths is unchanged by that.
    """
    if sets <= 0:
        return "rest", 0.0
    if sets <= target:
        return "trained", round(sets / target, 3)
    over = sets - target
    return "over", round(min(1.0, over / overshoot_span(target)), 3)


def summarise_entries(
    entries: list[WorkoutEntry], profile: TrainerProfile | None = None
) -> dict[str, dict]:
    """Aggregate ``entries`` into ``{muscle: {worked, sets, state, ...}}``.

    ``sets`` weights each set by how directly the movement trains the group: 3
    sets of bench press add 3 to *chest* (primary) and 1.5 to *triceps* and
    *shoulders* (secondary). Totals are therefore fractional.

    A movement outside :data:`app.exercises.VOLUME_CATEGORIES` is skipped
    entirely — it adds no sets and does not mark a group ``worked``, so a week
    of stretching leaves the body map grey rather than lit at zero intensity.

    Each group also carries its weekly ``target``, how many sets it is ``over``
    by, and the ``state``/``intensity`` pair the body map shades with.

    ``profile`` is the Phase 6 trainer setup, which decides what each group's
    target actually is; omitting it grades against the baseline convention,
    which is exactly what happened before that phase. Nothing else here changes
    with it — the aggregation is the same sets either way, and only the number
    they are measured against moves.
    """
    profile = profile or DEFAULT_PROFILE
    summary: dict[str, dict] = {
        muscle: {
            "muscle": muscle,
            "label": MUSCLE_LABELS[muscle],
            "worked": False,
            "sets": 0.0,
            "target": profile.target_for(muscle),
            "over": 0.0,
            "state": "rest",
            "intensity": 0.0,
            "exercises": [],
            "regions": [
                {
                    "region": region,
                    "label": REGION_LABELS[region],
                    "sets": 0.0,
                    "share": 0.0,
                    "neglected": False,
                }
                for region in regions_of(muscle)
            ],
            "region_sets": 0.0,
        }
        for muscle in MUSCLE_GROUPS
    }

    for entry in entries:
        exercise = get_exercise(entry.exercise_id)
        if exercise is None or not exercise.counts_toward_volume:
            continue

        for muscle in exercise.muscles:
            bucket = summary.get(muscle)
            if bucket is None:  # pragma: no cover - guards future exercise data
                continue
            weighted = entry.sets * exercise.weight_for(muscle)
            bucket["sets"] += weighted
            # Zero is reachable now: an entry of only warm-up sets has
            # entry.sets == 0, so weighted is 0 for every muscle it touches.
            # Before Phase 4 this branch was unreachable — CHECK (sets > 0)
            # guaranteed at least one real set — so don't drop the guard back
            # to an unconditional assignment; that would light the body map
            # for a group that received no volume.
            if weighted > 0:
                bucket["worked"] = True
                if entry.exercise_name not in bucket["exercises"]:
                    bucket["exercises"].append(entry.exercise_name)

            # A movement with no defensible emphasis inside the group — a
            # deadlift for the back — leaves this volume unattributed rather
            # than spreading it over regions it says nothing about. Movements
            # emphasising two regions split evenly between them.
            emphasised = regions_for(exercise.id, muscle)
            if not emphasised:
                continue
            per_region = weighted / len(emphasised)
            for region in bucket["regions"]:
                if region["region"] in emphasised:
                    region["sets"] += per_region
                    bucket["region_sets"] += per_region

    for bucket in summary.values():
        # Half-set weights land on .5 exactly, but round anyway so float error
        # never leaks a 12.499999999 into the payload.
        bucket["sets"] = round(bucket["sets"], 1)
        bucket["over"] = round(max(0.0, bucket["sets"] - bucket["target"]), 1)
        bucket["state"], bucket["intensity"] = grade(bucket["sets"], bucket["target"])
        _finish_regions(bucket)

    return summary


def _finish_regions(bucket: dict) -> None:
    """Fill in each region's share of its group, and whether it was left out.

    Shares are of the volume that could be *attributed* to a region, not of the
    group's total — so a week of deadlifts and pulldowns reports the lats' share
    of the pulldowns, and ``region_sets`` says how much of the back's volume that
    was. Reporting a share of the total instead would silently blame the deadlift
    for neglecting the mid back.
    """
    attributed = bucket["region_sets"]
    bucket["region_sets"] = round(attributed, 1)

    for region in bucket["regions"]:
        region["sets"] = round(region["sets"], 1)
        region["share"] = round(region["sets"] / attributed, 3) if attributed else 0.0
        region["neglected"] = (
            attributed >= REGION_NEGLECT_MIN_PARENT_SETS
            and region["share"] < REGION_NEGLECT_SHARE
        )


def weekly_summary(
    user_id: str,
    day: date,
    week_starts_on: int = 1,
    profile: TrainerProfile | None = None,
) -> dict:
    """Build the full payload backing the weekly summary page for ``day``.

    The resolved ``profile`` is echoed back beside the grading it produced, so a
    client renders the targets it was actually graded against rather than
    re-deriving them from a preference it holds locally.
    """
    profile = profile or DEFAULT_PROFILE
    start, end = week_bounds(day, week_starts_on)
    entries = list_entries(user_id, start, end)
    muscles = summarise_entries(entries, profile)

    per_day = {d.isoformat(): 0 for d in week_days(start)}
    for entry in entries:
        key = entry.entry_date.isoformat()
        if key in per_day:
            per_day[key] += entry.sets

    return {
        "week_start": start.isoformat(),
        "week_end": end.isoformat(),
        "profile": profile.to_dict(),
        "total_sets": sum(entry.sets for entry in entries),
        "total_entries": len(entries),
        "muscles": muscles,
        "muscles_worked": [m for m in MUSCLE_GROUPS if muscles[m]["worked"]],
        "muscles_at_target": [m for m in MUSCLE_GROUPS if muscles[m]["sets"] >= muscles[m]["target"]],
        "muscles_over": [m for m in MUSCLE_GROUPS if muscles[m]["state"] == "over"],
        # Flat list so the page can say "two regions were left out" without
        # walking every group. Order follows MUSCLE_GROUPS, then region order.
        "regions_neglected": [
            {"muscle": m, "region": r["region"], "label": r["label"]}
            for m in MUSCLE_GROUPS
            for r in muscles[m]["regions"]
            if r["neglected"]
        ],
        "sets_per_day": per_day,
        "entries": [entry.to_dict() for entry in entries],
    }

File: app/static/js/summary.js
/**
 * Weekly summary page — paints the body map and the per-muscle breakdown.
 *
 * Colour is a volume scale, not a yes/no: light green at one set deepening to
 * dark green at the group's weekly target, then light-to-dark red once the
 * target is passed. The server does the grading — each group arrives with a
 * `state` (`rest` | `trained` | `over`) and an `intensity` from 0 to 1 within
 * that state's ramp. All this module does is hand those to CSS.
 *
 * Set counts arrive fractional: a movement's primary muscles take the whole
 * set and its secondary muscles half, so `12.5 / 20` is a normal reading.
 *
 * The "inside the muscle" panel is the exception to all of the above: regions
 * are a *distribution*, never a grade. They take no target, no `--level` and
 * none of the ramp's colours — see `renderRegions` and docs/VOLUME_SCIENCE.md.
 */

import { fetchWeeklySummary } from "./api.js";
import {
  $, addDays, formatDate, formatSets, loadProfile, renderEntries, retargetLinks,
  saveProfile, syncUrlDate, toast,
} from "./ui.js";
import { initWeekStrip, setWeekStripDate } from "./weekstrip.js";

let anchorIso; // Any date inside the week being displayed.

/** `{scheme: [{key, label, muscles}, ...]}`, defined once in app/exercises.py. */
let schemeBuckets = {};

/** The chosen grouping. A view over the same rows — never a filter. */
let activeScheme;

/** The last week's per-muscle payload, so re-heading the list costs no request. */
let lastMuscles = null;

/**
 * Where the chosen split is remembered.
 *
 * Not in the URL: `?date=` is shared state that every page honours, while this
 * is a reading preference that should follow the reader rather than the link.
 */
const SCHEME_KEY = "bodyshop:summary-scheme";

/**
 * Apply a group's grade to one element: `--level` drives the colour mix, and
 * the classes pick which ramp CSS mixes along.
 */
function applyGrade(element, info) {
  element.style.setProperty("--level", info ? info.intensity : 0);
  element.classList.toggle("is-worked", Boolean(info && info.worked));
  element.classList.toggle("is-over", Boolean(info && info.state === "over"));
}

/** Describe a group's weekly volume for the region's tooltip. */
function describe(info, fallbackLabel) {
  if (!info) return fallbackLabel;
  if (!info.worked) return `${info.label}: no sets this week (target ${info.target})`;

  const sets = `${formatSets(info.sets)} of ${info.target} sets`;
  const detail = info.exercises.join(", ");
  return info.state === "over"
    ? `${info.label}: ${sets} — ${formatSets(info.over)} over target (${detail})`
    : `${info.label}: ${sets} (${detail})`;
}

/** Shade every SVG region according to its group's weekly volume. */
function paintBody(muscles) {
  document.querySelectorAll(".muscle").forEach((region) => {
    const info = muscles[region.dataset.muscle];
    applyGrade(region, info);

    const title = region.querySelector("title");
    if (title) title.textContent = describe(info, region.dataset.muscle);
  });

  // Arms the staggered head-to-toe sweep in CSS. Set after the grades so the
  // delays apply to this transition rather than to the page's first paint,
  // which would hold an already-grey map grey for another half second.
  document.querySelectorAll(".body-muscles").forEach((group) => {
    group.classList.add("is-graded");
  });
}

/** Fill the breakdown list, each bar scaled against its target. */
function renderBreakdown(muscles) {
  document.querySelectorAll(".muscle-row").forEach((row) => {
    const info = muscles[row.dataset.muscle];
    if (!info) return;
    applyGrade(row, info);

    // The bar tops out at the target; overshoot shows as colour, not length.
    const progress = Math.min(1, info.sets / info.target);
    row.querySelector(".muscle-bar-fill").style.width = `${progress * 100}%`;
    row.querySelector(".muscle-sets").textContent =
      `${formatSets(info.sets)} / ${info.target}`;
    row.title = describe(info, row.dataset.muscle);
  });

  renderBucketTotals(muscles);
}

/**
 * Total each visible bucket.
 *
 * Sets only — no bucket target. Summing twelve targets into a "push target"
 * would put a number on screen that nobody has studied, which is the same
 * reason regions are not graded (docs/VOLUME_SCIENCE.md).
 */
function renderBucketTotals(muscles) {
  (schemeBuckets[activeScheme] || []).forEach((bucket) => {
    const heading = document.querySelector(
      `.muscle-bucket[data-scheme="${activeScheme}"][data-bucket="${bucket.key}"]`,
    );
    if (!heading) return;

    const sets = bucket.muscles.reduce(
      (total, muscle) => total + (muscles[muscle] ? muscles[muscle].sets : 0), 0,
    );
    heading.querySelector(".muscle-bucket-sets").textContent =
      sets ? `${formatSets(sets)} sets` : "nothing logged";
  });
}

/**
 * Re-head the breakdown into `key`.
 *
 * Nodes are *moved* rather than given a CSS `order`, so the reading order a
 * screen reader follows matches the order on screen. Every scheme files all
 * twelve rows, so nothing can be hidden by switching view — only regrouped.
 */
function applyScheme(key) {
  const buckets = schemeBuckets[key];
  if (!buckets) return;
  activeScheme = key;

  const list = $("#muscle-list");
  const rows = new Map(
    Array.from(list.querySelectorAll(".muscle-row")).map((row) => [row.dataset.muscle, row]),
  );

  document.querySelectorAll(".muscle-bucket").forEach((heading) => {
    heading.hidden = heading.dataset.scheme !== key;
  });

  buckets.forEach((bucket) => {
    const heading = list.querySelector(
      `.muscle-bucket[data-scheme="${key}"][data-bucket="${bucket.key}"]`,
    );
    if (heading) list.append(heading);
    bucket.muscles.forEach((muscle) => {
      const row = rows.get(muscle);
      if (row) list.append(row);
    });
  });

  // Park the other schemes' headings at the end. They are hidden either way,
  // but leaving them in front would make one of them `:first-child` and give
  // the first visible heading a rule above it.
  list.querySelectorAll(".muscle-bucket[hidden]").forEach((h) => list.append(h));

  const select = $("#scheme-select");
  select.value = key;
  $("#scheme-note").textContent = select.selectedOptions[0]?.dataset.note || "";

  try {
    localStorage.setItem(SCHEME_KEY, key);
  } catch {
    // Private browsing, or storage disabled. The choice just won't persist.
  }
}

function onSchemeChange(event) {
  applyScheme(event.target.value);
  // The week already loaded; the new headings just need totalling. No refetch —
  // switching view is not new data.
  if (lastMuscles) renderBucketTotals(lastMuscles);
}

/**
 * Fill the "inside the muscle" panel: where each subdivided group's work landed.
 *
 * Regions are **not** graded. They get no target, no `--level` and none of the
 * volume ramp's colours, because no evidence establishes how many sets a muscle
 * head needs (docs/VOLUME_SCIENCE.md). The bar length is the region's share of
 * the volume the server could place inside the group, and the only judgement
 * shown is `neglected`, which the server decides.
 */
/**
 * Say which sets the bars below are a share *of*, and why it can be fewer than
 * the group's total.
 *
 * Always the same shape — "based on …" — because the two cases differ only in
 * whether every set could be placed, and phrasing them differently made them
 * look like unrelated readings.
 */
function setBasis(element, info) {
  const total = formatSets(info.sets);
  const placed = formatSets(info.region_sets);
  const missed = formatSets(info.sets - info.region_sets);

  if (!info.worked) {
    element.textContent = "not trained this week";
    element.title = "";
  } else if (!info.region_sets) {
    element.textContent = "no split to show";
    element.title =
      `All ${total} sets came from movements that do not favour one region of `
      + "this group, so there is nothing to divide up.";
  } else if (info.region_sets < info.sets) {
    element.textContent = `based on ${placed} of ${total} sets`;
    element.title =
      `${missed} of this group's ${total} sets came from movements that do not `
      + "favour one region — a deadlift trains the back without telling you "
      + "lats or mid back — so they count for the group but not for a bar here.";
  } else {
    element.textContent = `based on all ${total} sets`;
    element.title = `Every one of this group's ${total} sets favoured a region.`;
  }
}

function renderRegions(muscles) {
  document.querySelectorAll(".region-group").forEach((group) => {
    const info = muscles[group.dataset.muscle];
    if (!info) return;

    setBasis(group.querySelector('[data-role="basis"]'), info);

    const byRegion = new Map(info.regions.map((r) => [r.region, r]));
    group.querySelectorAll(".region-bar").forEach((bar) => {
      const region = byRegion.get(bar.dataset.region);
      if (!region) return;

      bar.querySelector(".region-bar-fill").style.width = `${region.share * 100}%`;
      bar.querySelector(".region-bar-value").textContent =
        info.region_sets ? `${Math.round(region.share * 100)}%` : "—";
      bar.classList.toggle("is-neglected", region.neglected);
      bar.title = region.neglected
        ? `${region.label}: ${formatSets(region.sets)} sets — thin next to the rest of this group`
        : `${region.label}: ${formatSets(region.sets)} sets`;
    });
  });
}

/**
 * One readout cell: a figure over the thing it counts.
 *
 * `over` marks the only cell allowed the ramp's colour — it reports the same
 * fact the map is already showing in red, so tinting it agrees with the map
 * rather than inventing a second scale.
 */
function readoutCell(value, label, { over = false } = {}) {
  const cell = document.createElement("div");
  cell.className = `readout-cell${over ? " is-over" : ""}`;

  const figure = document.createElement("dd");
  figure.className = "readout-value";
  figure.textContent = value;

  const name = document.createElement("dt");
  name.className = "readout-label type-label";
  name.textContent = label;

  // <dt> before <dd> in the DOM, so the pair reads correctly; CSS puts the
  // figure on top.
  cell.append(name, figure);
  return cell;
}

function renderHeader(summary) {
  const start = formatDate(summary.week_start, { month: "short", day: "numeric" });
  const end = formatDate(summary.week_end, { month: "short", day: "numeric", year: "numeric" });
  $("#week-label").textContent = `${start} – ${end}`;

  const worked = summary.muscles_worked.length;
  const total = Object.keys(summary.muscles).length;
  const thin = summary.regions_neglected.length;

  const meta = $("#week-meta");
  meta.textContent = "";
  meta.append(
    readoutCell(summary.total_sets, "Sets"),
    readoutCell(`${worked}/${total}`, "Trained"),
    readoutCell(summary.muscles_at_target.length, "At target"),
    summary.muscles_over.length
      ? readoutCell(summary.muscles_over.length, "Over", { over: true })
      : readoutCell(thin, thin === 1 ? "Region thin" : "Regions thin"),
  );
}

// ---- Trainer setup (Phase 6) ----------------------------------------------

/**
 * Read the three controls into the shape `ui.js` stores and `api.js` sends.
 *
 * Deliberately unvalidated beyond `Number`: `resolve_profile` on the server
 * clamps to the same bounds the inputs carry, and duplicating that here would
 * be a second implementation of one rule.
 */
function readSetup() {
  return {
    experience: $("#experience-select").value,
    sessions_per_week: Number($("#sessions-input").value),
    minutes_per_session: Number($("#minutes-input").value),
  };
}

/** Put a stored setup back into the controls on first paint. */
function fillSetup(profile) {
  $("#experience-select").value = profile.experience;
  $("#sessions-input").value = profile.sessions_per_week;
  $("#minutes-input").value = profile.minutes_per_session;
  renderBlurb();
}

function renderBlurb() {
  const option = $("#experience-select").selectedOptions[0];
  $("#experience-blurb").textContent = (option && option.dataset.blurb) || " ";
}

/**
 * Say what the setup resolved to, in words.
 *
 * **One number per size of group, never a range** — printing "aim for 10–20"
 * would invite reading the top of it as the goal, which is the product-voice
 * rule in docs/VOLUME_SCIENCE.md §4. And the two figures come from the server's
 * own `targets`, not from re-applying the multiplier here, so the sentence
 * cannot disagree with the bars underneath it.
 */
function renderSetupEffect(profile) {
  const large = profile.targets.chest;
  const small = profile.targets.abs;

  $("#setup-summary").textContent =
    `${profile.sessions_per_week} × ${profile.minutes_per_session} min`;

  const reason = profile.limited_by === "plan"
    ? "Your week is what is setting these — more time, or another session, "
      + "raises them until they reach what your experience asks for."
    : "Your experience level is what is setting these; the time you have "
      + "covers them.";

  $("#setup-effect").textContent =
    `Large groups are covered at ${large} sets a week, small ones at ${small}. `
    + reason;
}

async function onSetupChange() {
  saveProfile(readSetup());
  renderBlurb();
  // Targets are graded server-side, so a new setup is a new request rather than
  // a re-render: the states, the ramp positions and the readouts all move with
  // it, and recomputing any of them here would be a second grader.
  await load();
}

async function load() {
  try {
    const summary = await fetchWeeklySummary(anchorIso);
    lastMuscles = summary.muscles;
    // Echoed back by the server, so the controls settle on what was actually
    // used — a value clamped out of range corrects itself on screen instead of
    // sitting there disagreeing with the bars it produced.
    if (summary.profile) {
      fillSetup(summary.profile);
      renderSetupEffect(summary.profile);
    }
    renderHeader(summary);
    paintBody(summary.muscles);
    renderBreakdown(summary.muscles);
    renderRegions(summary.muscles);
    renderEntries($("#week-entries"), summary.entries, {
      showDate: true,
      emptyMessage: "No workouts logged this week yet.",
    });
  } catch (err) {
    toast(err.message, "error");
  }
}

/**
 * Move the page to `iso` — the one place the anchor date changes.
 *
 * Stepping a week and clicking a day in the calendar strip are the same act,
 * so they go through here: `?date=` is shared state every page honours, and the
 * strip has to re-anchor with the summary or the two disagree about which week
 * is on screen.
 */
async function goToDate(iso) {
  anchorIso = iso;
  syncUrlDate(anchorIso);
  retargetLinks(
    Array.from(document.querySelectorAll(".nav-link, .tab-link, .shelf")), anchorIso,
  );
  await Promise.all([load(), setWeekStripDate(anchorIso)]);
}

function shiftWeek(days) {
  goToDate(addDays(anchorIso, days));
}

/** The stored split, if it is still one the server offers. */
function storedScheme() {
  try {
    const stored = localStorage.getItem(SCHEME_KEY);
    return stored && schemeBuckets[stored] ? stored : null;
  } catch {
    return null;
  }
}

/**
 * Boot the summary page.
 * @param {string} initialIso - Any date inside the week to show.
 * @param {Object} buckets - Grouping schemes from `app/exercises.py`, keyed by
 *   scheme: `{push_pull_legs: [{key, label, muscles}, ...], ...}`.
 */
export async function initSummary(initialIso, buckets = {}) {
  anchorIso = initialIso;
  schemeBuckets = buckets;
  $("#prev-week").addEventListener("click", () => shiftWeek(-7));
  $("#next-week").addEventListener("click", () => shiftWeek(7));
  $("#scheme-select").addEventListener("change", onSchemeChange);

  // Fill the setup controls *before* the first fetch, so the request carries
  // the stored profile rather than the markup's defaults and the page never
  // paints one set of targets and then replaces it with another.
  fillSetup(loadProfile());
  ["#experience-select", "#sessions-input", "#minutes-input"].forEach((selector) => {
    $(selector).addEventListener("change", onSetupChange);
  });

  applyScheme(storedScheme() || $("#scheme-select").value);
  await load();
  // After the week, not before: the strip is a supporting reading and the body
  // map is what the page is for, so it does not delay the thing being read.
  await initWeekStrip(anchorIso, goToDate);
}
